src/overlay/SurveyManager.cpp

```diff
@@ -77,10 +144,156 @@ SurveyManager::stopSurvey()
+        {
+            return;
+        }
+    }
+
+    auto onSuccessValidation = [&]() -> bool {
+        // Check signature
+        return dropPeerIfSigInvalid(startRecording.surveyorID,
+                                    signedStartRecording.signature,
+                                    xdr::xdr_to_opaque(startRecording), peer);
+    };
+
+    if (!mMessageLimiter.validateStartSurveyRecording(startRecording,
+                                                      onSuccessValidation))
+    {
+        return;
+    }
+
+    if (!mApp.getSurveyDataManager().startSurveyRecording(
+            startRecording, mApp.getOverlayManager().getAuthenticatedPeers()))
+    {
+        return;
+    }
+
+    if (peer)
+    {
+        mApp.getOverlayManager().recvFloodedMsg(msg, peer);
+    }
+
+    broadcast(msg);
+}
+
+void
+SurveyManager::broadcastStopSurveyRecording(uint32_t nonce)
+{
+    StellarMessage newMsg;
+    newMsg.type(TIME_SLICED_SURVEY_STOP_RECORDING);
+    auto& signedStopRecording =
+        newMsg.signedTimeSlicedSurveyStopRecordingMessage();
+    auto& stopRecording = signedStopRecording.stopRecording;
+
+    stopRecording.surveyorID = mApp.getConfig().NODE_SEED.getPublicKey();
+    stopRecording.nonce = nonce;
+    stopRecording.ledgerNum = mApp.getHerder().trackingConsensusLedgerIndex();
+
+    auto sigBody = xdr::xdr_to_opaque(stopRecording);
+    signedStopRecording.signature = mApp.getConfig().NODE_SEED.sign(sigBody);
+
+    relayStopSurveyRecording(newMsg, nullptr);
+}
+
+void
+SurveyManager::relayStopSurveyRecording(StellarMessage const& msg,
+                                        Peer::pointer peer)
+{
+    releaseAssert(msg.type() == TIME_SLICED_SURVEY_STOP_RECORDING);
+    auto const& signedStopRecording =
+        msg.signedTimeSlicedSurveyStopRecordingMessage();
+    auto const& stopRecording = signedStopRecording.stopRecording;
+
+    auto surveyorIsSelf =
+        stopRecording.surveyorID == mApp.getConfig().NODE_SEED.getPublicKey();
+    if (!surveyorIsSelf)
+    {
+        releaseAssert(peer);
+
+        if (!surveyorPermitted(stopRecording.surveyorID))
+        {
+            return;
+        }
+    }
+
+    auto onSuccessValidation = [&]() -> bool {
+        // Check signature
+        return dropPeerIfSigInvalid(stopRecording.surveyorID,
+                                    signedStopRecording.signature,
+                                    xdr::xdr_to_opaque(stopRecording), peer);
+    };
+
+    if (!mMessageLimiter.validateStopSurveyRecording(stopRecording,
+                                                     onSuccessValidation))
+    {
+        return;
+    }
+
+    if (!mApp.getSurveyDataManager().stopSurveyRecording(
+            stopRecording, mApp.getOverlayManager().getAuthenticatedPeers(),
+            mApp.getConfig()))
+    {
+        return;
+    }
+
+    if (peer)
+    {
+        mApp.getOverlayManager().recvFloodedMsg(msg, peer);
+    }
+
+    broadcast(msg);
```

Comment:
On the testing front, I wonder if it'd be easy to add a basic smoke test in supercluster, where we spin up a network of mixed core versions, and trigger survey. We probably don't need to validate results, as that's something we can do with unit tests, but we'll get an end-to-end survey run with mixed versions. If this is too involved, we don't have to do it this time around, and instead try a small network of cores locally, for example (but that's more of a one-off).

Reply:
Adding a note here for myself: I'll need to add a config option to lower the timeouts so we can hit them in the supercluster run itself (something like `ARTIFICIALLY_LOWER_SURVEY_TIMEOUTS_FOR_TESTING`)